README.md
# php-byrobot

[![Latest Version on Packagist][ico-version]][link-packagist]
[![Software License][ico-license]](LICENSE.md)
[![Build Status][ico-travis]][link-travis]
[![Coverage Status][ico-scrutinizer]][link-scrutinizer]
[![Quality Score][ico-code-quality]][link-code-quality]
[![Total Downloads][ico-downloads]][link-downloads]

百应智能外呼系统 API

## Structure

If any of the following are applicable to your project, then the directory structure should follow industry best practices by being named the following.

```
src/
tests/
```

## Install

Via Composer

```bash
$ composer require yiranzai/byrobot
```

## Usage

### 初始化

```php
$robot = Yiranzai\Byrobot\Byrobot::init(
    [
        'appKey' => 'YOU_APP_KEY',
        'appSecret' => 'YOU_APP_SECRET'
    ]);
```

or

```php
$robot = Yiranzai\Byrobot\Byrobot::init();
$robot->key(YOU_APP_KEY)->secret(YOU_APP_SECRET);
```

### 使用

- `$robot->getCompanys()` // 获取绑定公司列表接口
- `$robot->getPhones()` // 获取公司的主叫电话列表接口
- `$robot->getRobots()` // 获取公司的机器人话术列表接口
- `$robot->addBlackList()` // 添加单个黑名单到公司默认黑名单组接口
- `$robot->statistics()` // 获取公司AI坐席概况接口
- `$robot->createTask()` // 创建任务接口
- `$robot->start()` // 启动任务接口
- `$robot->pause()` // 暂停任务接口
- `$robot->stop()` // 停止任务接口
- `$robot->delete()` // 删除任务
- `$robot->importTaskCustomer()` // 向任务中导入客户接口
- `$robot->update()` // 修改任务
- `$robot->call()` // 单次电话外呼
- `$robot->singleCallByMobile()` // 根据手机号进行单次电话外呼
- `$robot->getTasks()` // 获取任务列表接口
- `$robot->getTaskDetail()` // 获取任务详情接口
- `$robot->queryDoneTaskPhones()` // 获取已经完成任务电话号码接口
- `$robot->notDialedCustomerList()` // 获取任务未开始的电话列表
- `$robot->phoneLogInfo()` // 获取一个通话的详情接口

## Change log

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Testing

```bash
$ composer test
```

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) and [CODE_OF_CONDUCT](CODE_OF_CONDUCT.md) for details.

## Security

If you discover any security related issues, please email wuqingdzx@gmail.com instead of using the issue tracker.

## Credits

-   [yiranzai][link-author]
-   [All Contributors][link-contributors]

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/yiranzai/byrobot.svg?style=flat-square
[ico-license]: https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/yiranzai/php-byrobot/master.svg?style=flat-square
[ico-scrutinizer]: https://img.shields.io/scrutinizer/coverage/g/yiranzai/php-byrobot.svg?style=flat-square
[ico-code-quality]: https://img.shields.io/scrutinizer/g/yiranzai/php-byrobot.svg?style=flat-square
[ico-downloads]: https://img.shields.io/packagist/dt/yiranzai/byrobot.svg?style=flat-square
[link-packagist]: https://packagist.org/packages/yiranzai/byrobot
[link-travis]: https://travis-ci.org/yiranzai/php-byrobot
[link-scrutinizer]: https://scrutinizer-ci.com/g/yiranzai/php-byrobot/code-structure
[link-code-quality]: https://scrutinizer-ci.com/g/yiranzai/php-byrobot
[link-downloads]: https://packagist.org/packages/yiranzai/byrobot
[link-author]: https://github.com/yiranzai
[link-contributors]: ../../contributors


CHANGELOG.md
# Changelog

All notable changes to `byrobot` will be documented in this file.

Updates should follow the [Keep a CHANGELOG](http://keepachangelog.com/) principles.


## v0.2 - 2019-03-22

简单修复

### Fixed

- `$robot->getTasks()` // pageSize限制有错误修正为 `(2<<30) -1`
- `$robot->queryDoneTaskPhones()` // pageSize限制有错误，修正为 `50`

## v0.1 - 2019-03-21
初步完成
### Added

- `$robot->getCompanys()` // 获取绑定公司列表接口
- `$robot->getPhones()` // 获取公司的主叫电话列表接口
- `$robot->getRobots()` // 获取公司的机器人话术列表接口
- `$robot->addBlackList()` // 添加单个黑名单到公司默认黑名单组接口
- `$robot->statistics()` // 获取公司AI坐席概况接口
- `$robot->createTask()` // 创建任务接口
- `$robot->start()` // 启动任务接口
- `$robot->pause()` // 暂停任务接口
- `$robot->stop()` // 停止任务接口
- `$robot->delete()` // 删除任务
- `$robot->importTaskCustomer()` // 向任务中导入客户接口
- `$robot->update()` // 修改任务
- `$robot->call()` // 单次电话外呼
- `$robot->singleCallByMobile()` // 根据手机号进行单次电话外呼
- `$robot->getTasks()` // 获取任务列表接口
- `$robot->getTaskDetail()` // 获取任务详情接口
- `$robot->queryDoneTaskPhones()` // 获取已经完成任务电话号码接口
- `$robot->notDialedCustomerList()` // 获取任务未开始的电话列表
- `$robot->phoneLogInfo()` // 获取一个通话的详情接口



tests/ByrobotTest.php
<?php
/**
 * Created by PhpStorm.
 * User: yiranzai
 * Date: 19-3-22
 * Time: 下午3:15
 */

namespace Yiranzai\Byrobot\Tests;

use PHPUnit\Framework\TestCase;
use Yiranzai\Byrobot\Byrobot;

/**
 * Class ByrobotTest
 * @package Yiranzai\Byrobot\Tests
 */
class ByrobotTest extends TestCase
{
    /**
     *
     */
    public function testByrobotOne()
    {
        $robot = Byrobot::init(['appKey' => 'test', 'appSecret' => 'test']);
        $this->assertSame($robot, Byrobot::init());
        $this->assertSame($robot, Byrobot::init());
    }

    /**
     *
     */
    public function testByrobotTwo()
    {
        $robot  = Byrobot::init(['appKey' => 'test', 'appSecret' => 'test']);
        $msgOne = $msgTwo = null;
        try {
            $robot->getCompanys();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->getCompanys([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->getPhones();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->getPhones([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->getRobots();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->getRobots([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->addBlackList();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->addBlackList([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->statistics();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->statistics([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->createTask();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->createTask([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->start();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->start([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->pause();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->pause([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->stop();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->stop([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->delete();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->delete([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->importTaskCustomer();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->importTaskCustomer([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->update();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->update([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->call();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->call([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->singleCallByMobile();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->singleCallByMobile([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->getTasks();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->getTasks([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->getTaskDetail();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->getTaskDetail([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->queryDoneTaskPhones();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->queryDoneTaskPhones([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->notDialedCustomerList();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->notDialedCustomerList([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
        try {
            $robot->phoneLogInfo();
        } catch (\Exception $e) {
            $msgOne = $e->getMessage();
        }
        try {
            $robot->phoneLogInfo([], true);
        } catch (\Exception $e) {
            $msgTwo = $e->getMessage();
        }
        $this->assertSame($msgOne, $msgTwo);
    }

    /**
     * @throws \Exception
     */
    public function testByrobotExceptionOne()
    {
        $this->expectException(\Exception::class);
        $robot = Byrobot::init();
        $robot->getCompanys();
    }

    /**
     * @throws \Exception
     */
    public function testByrobotExceptionTwo()
    {
        $this->expectException(\Exception::class);
        $robot = Byrobot::init(['appKey' => 'test', 'appSecret' => 'test']);
        $robot->getTasks(['pageSize' => (2 << 32)]);
    }

    /**
     * @throws \Exception
     */
    public function testByrobotExceptionThree()
    {
        $this->expectException(\Exception::class);
        $robot = Byrobot::init(['appKey' => 'test', 'appSecret' => 'test']);
        $robot->queryDoneTaskPhones(['pageSize' => 51]);
    }
}


src/Byrobot.php
<?php
/**
 * Created by PhpStorm.
 * User: yiranzai
 * Date: 19-3-19
 * Time: 上午10:32
 */

namespace Yiranzai\Byrobot;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;
use Yiranzai\Byrobot\Traits\Base;
use Yiranzai\Byrobot\Traits\Byrobot as ByrobotTrait;
use Yiranzai\Tools\Date;
use Yiranzai\Tools\Tools;

/**
 * Class Byrobot
 * @package Yiranzai\Byrobot
 */
final class Byrobot
{
    use ByrobotTrait, Base;
    /**
     * @var array
     */
    private static $guarded = ['url', 'client', 'guarded'];
    /**
     * @var Client
     */
    private $client;

    /**
     * Byrobot constructor.
     * @param array $config
     */
    private function __construct(array $config = [])
    {

        foreach ($config as $key => $value) {
            if (in_array($key, self::$guarded, true)) {
                continue;
            }
            self::$$key = $value;
        }
    }

    /**
     * 获取绑定公司列表接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function getCompanys($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * @param       $str
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    private function send($str, $param = [], $isArr = false)
    {
        if (!$this->check()) {
            throw  new \RuntimeException('核心参数不足');
        }
        ['url' => $url, 'method' => $method, 'desc' => $desc] = $this->build($str);
        [$data, $msg] = $this->request($url, $method, $param, $isArr);
        if ($msg) {
            throw  new \RuntimeException($url . ' | ' . $desc . ' | ' . $msg);
        }
        return $data;
    }

    /**
     * @return bool
     */
    private function check(): bool
    {
        return self::$appKey !== null && self::$appSecret !== null;
    }

    /**
     * @param $str
     * @return array
     * @throws \Exception
     */
    protected function build($str): array
    {
        if (!isset(self::$url[$str])) {
            throw  new \RuntimeException($str . ' api not found');
        }
        $data        = self::$url[$str];
        $data['url'] = self::$baseUrl .
            str_replace('{apiVersion}', self::$apiVersion, $data['url']);
        return $data;
    }

    /**
     * @param string $url
     * @param string $method
     * @param array  $param
     * @param bool   $isArr
     * @return array
     */
    protected function request($url, $method, $param = [], $isArr = false): array
    {
        $msg = $json = null;
        try {
            $date  = Date::toCarbon()->toRfc7231String();
            $sign  = $this->generateSign($date);
            $query = [];

            $method = strtoupper($method);

            if ($method === 'GET') {
                $query['query'] = $param;
            }

            if ($method === 'POST') {
                $query['json'] = $param;
            }

            $query['headers'] = [
                'datetime' => $date,
                'appkey'   => self::$appKey,
                'sign'     => $sign
            ];
            $response         = $this->getClient()->request($method, $url, $query);
            //状态码不是200直接抛出异常
            if (200 !== $response->getStatusCode()) {
                throw new \RuntimeException('请求出错！'
                    . Tools::arrGet(self::$errorCode, $response->getStatusCode(), ''));
            }

            //返回的内容不是json,抛出异常
            $json = json_decode($response->getBody(), $isArr);
            if (JSON_ERROR_NONE !== json_last_error()) {
                $json = null;
                throw new \RuntimeException('请求结果异常！' . Tools::iteratorGet($json, 'resultMsg', ''));
            }

            if (200 !== $code = (int)Tools::iteratorGet($json, 'code')) {
                $json = null;
                throw new \RuntimeException('请求返回出错！' . Tools::arrGet(self::$errorCode, $code, ''));
            }
            $json = Tools::iteratorGet($json, 'data');
        } catch (\Throwable $e) {
            $msg = $e->getMessage();
        } catch (GuzzleException $e) {
            $msg = $e->getMessage();
        }
        return [$json, $msg];
    }

    /**
     * 加密
     * 规则：GMT时间 + "\n" + appKey
     *
     * @param String $date 时间：Thu, 13 Dec 2018 01:27:17 GMT
     * @return string
     */
    private function generateSign(string $date): string
    {
        if (function_exists('hash_hmac')) {
            return base64_encode(hash_hmac('sha1', self::$appKey . "\n" . $date, self::$appSecret, true));
        }
        return null;
    }

    /**
     * @return Client
     */
    protected function getClient(): Client
    {
        if (null === $this->client) {
            $this->client = new Client();
        }
        return $this->client;
    }

    /**
     * 获取公司的主叫电话列表接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function getPhones($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 获取公司的机器人话术列表接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function getRobots($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 添加单个黑名单到公司默认黑名单组接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function addBlackList($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 获取公司AI坐席概况接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function statistics($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 创建任务接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function createTask($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 启动任务接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function start($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 暂停任务接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function pause($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 停止任务接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function stop($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 删除任务
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function delete($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 向任务中导入客户接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function importTaskCustomer($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 修改任务
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function update($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 单次电话外呼
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function call($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 根据手机号进行单次电话外呼
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function singleCallByMobile($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 获取任务列表接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function getTasks($param = [], $isArr = false)
    {
        if (isset($param['pageSize']) && $param['pageSize'] > self::$maxTaskPageSize) {
            throw  new \RuntimeException('pageSize 不能大于' . self::$maxTaskPageSize);
        }
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 获取任务详情接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function getTaskDetail($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 获取已经完成任务电话号码接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function queryDoneTaskPhones($param = [], $isArr = false)
    {
        if (isset($param['pageSize']) && $param['pageSize'] > self::$maxPhoneLogPageSize) {
            throw  new \RuntimeException('pageSize 不能大于' . self::$maxPhoneLogPageSize);
        }
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 获取任务未开始的电话列表
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function notDialedCustomerList($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * 获取一个通话的详情接口
     *
     * @param array $param
     * @param bool  $isArr
     * @return array|object|null
     * @throws \Exception
     */
    public function phoneLogInfo($param = [], $isArr = false)
    {
        return $this->send(__FUNCTION__, $param, $isArr);
    }

    /**
     * prevent from being unserialized (which would create a second instance of it)
     */
    public function __wakeup()
    {
    }


    /**
     * prevent the instance from being cloned (which would create a second instance of it)
     */
    private function __clone()
    {
    }
}


src/Traits/Base.php
<?php
/**
 * Created by PhpStorm.
 * User: yiranzai
 * Date: 19-3-22
 * Time: 下午2:19
 */

namespace Yiranzai\Byrobot\Traits;

/**
 * Trait Base
 * @package Yiranzai\Byrobot\Traits
 */
trait Base
{
    /**
     * @var string
     */
    private static $appKey;
    /**
     * @var string
     */
    private static $appSecret;
    /**
     * 接口域名
     * @var string
     */
    private static $baseUrl = 'https://api.byrobot.cn';
    /**
     * 接口版本
     * @var string
     */
    private static $apiVersion = 'v1';

    /**
     * @var \Yiranzai\Byrobot\Byrobot
     */
    private static $robot;

    /**
     * @var array
     */
    public static $url = [
        'getCompanys'           => [
            'url'    => '/openapi/{apiVersion}/company/getCompanys',
            'method' => 'GET',
            'desc'   => '获取绑定公司列表接口',
        ],
        'getPhones'             => [
            'url'    => '/openapi/{apiVersion}/company/getPhones',
            'method' => 'GET',
            'desc'   => '获取公司的主叫电话列表接口',
        ],
        'getRobots'             => [
            'url'    => '/openapi/{apiVersion}/company/getRobots',
            'method' => 'GET',
            'desc'   => '获取公司的机器人话术列表接口',
        ],
        'addBlackList'          => [
            'url'    => '/openapi/{apiVersion}/company/addBlackList',
            'method' => 'POST',
            'desc'   => '添加单个黑名单到公司默认黑名单组接口',
        ],
        'statistics'            => [
            'url'    => '/openapi/{apiVersion}/company/seat/statistics',
            'method' => 'GET',
            'desc'   => '获取公司AI坐席概况接口',
        ],
        'createTask'            => [
            'url'    => '/openapi/{apiVersion}/task/createTask',
            'method' => 'POST',
            'desc'   => '创建任务接口',
        ],
        'start'                 => [
            'url'    => '/openapi/{apiVersion}/task/start',
            'method' => 'POST',
            'desc'   => '启动任务接口',
        ],
        'pause'                 => [
            'url'    => '/openapi/{apiVersion}/task/pause',
            'method' => 'POST',
            'desc'   => '暂停任务接口',
        ],
        'stop'                  => [
            'url'    => '/openapi/{apiVersion}/task/stop',
            'method' => 'POST',
            'desc'   => '停止任务接口',
        ],
        'delete'                => [
            'url'    => '/openapi/{apiVersion}/task/delete',
            'method' => 'POST',
            'desc'   => '删除任务',
        ],
        'importTaskCustomer'    => [
            'url'    => '/openapi/{apiVersion}/task/importTaskCustomer',
            'method' => 'POST',
            'desc'   => '向任务中导入客户接口',
        ],
        'update'                => [
            'url'    => '/openapi/{apiVersion}/task/update',
            'method' => 'POST',
            'desc'   => '修改任务',
        ],
        'call'                  => [
            'url'    => '/openapi/{apiVersion}/task/call',
            'method' => 'POST',
            'desc'   => '单次电话外呼',
        ],
        'singleCallByMobile'    => [
            'url'    => '/openapi/{apiVersion}/task/singleCallByMobile',
            'method' => 'POST',
            'desc'   => '根据手机号进行单次电话外呼',
        ],
        'getTasks'              => [
            'url'    => '/openapi/{apiVersion}/task/getTasks',
            'method' => 'GET',
            'desc'   => '获取任务列表接口',
        ],
        'getTaskDetail'         => [
            'url'    => '/openapi/{apiVersion}/task/getTaskDetail',
            'method' => 'GET',
            'desc'   => '获取任务详情接口',
        ],
        'queryDoneTaskPhones'   => [
            'url'    => '/openapi/{apiVersion}/task/queryDoneTaskPhones',
            'method' => 'POST',
            'desc'   => '获取已经完成任务电话号码接口',
        ],
        'notDialedCustomerList' => [
            'url'    => '/openapi/{apiVersion}/task/notDialedCustomerList',
            'method' => 'POST',
            'desc'   => '获取任务未开始的电话列表',
        ],
        'phoneLogInfo'          => [
            'url'    => '/openapi/{apiVersion}/task/phoneLogInfo',
            'method' => 'GET',
            'desc'   => '获取一个通话的详情接口',
        ],
    ];

    /**
     * @param $key
     * @return $this
     */
    public function key($key): self
    {
        self::$appKey = $key;
        return $this;
    }

    /**
     * @param $secret
     * @return $this
     */
    public function secret($secret): self
    {
        self::$appSecret = $secret;
        return $this;
    }

    /**
     * @param $version
     * @return $this
     */
    public function version($version): self
    {
        self::$apiVersion = $version;
        return $this;
    }

    /**
     * @param $url
     * @return $this
     */
    public function url($url): self
    {
        self::$baseUrl = $url;
        return $this;
    }

    /**
     * 获取百应机器人实例
     *
     * @param array $config
     * @return \Yiranzai\Byrobot\Byrobot
     */
    public static function init(array $config = []): \Yiranzai\Byrobot\Byrobot
    {
        if (null === static::$robot) {
            static::$robot = new static($config);
        }
        return static::$robot;
    }
}
